x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/ReservedRoleMappingAction.java

```diff
@@ -0,0 +1,141 @@
+    public static final String NAME = "role_mappings";
+
+    private final NativeRoleMappingStore roleMappingStore;
+
+    /**
+     * Creates a ReservedRoleMappingAction
+     *
+     * @param roleMappingStore requires {@link NativeRoleMappingStore} for storing/deleting the mappings
+     */
+    public ReservedRoleMappingAction(NativeRoleMappingStore roleMappingStore) {
+        this.roleMappingStore = roleMappingStore;
+    }
+
+    @Override
+    public String name() {
+        return NAME;
+    }
+
+    @SuppressWarnings("unchecked")
+    public Collection<PutRoleMappingRequest> prepare(Object input) {
+        List<ExpressionRoleMapping> roleMappings = (List<ExpressionRoleMapping>) input;
+        List<PutRoleMappingRequest> requests = roleMappings.stream().map(rm -> PutRoleMappingRequest.fromMapping(rm)).toList();
+
+        var exceptions = new ArrayList<String>();
+        for (var request : requests) {
+            var exception = request.validate();
+            if (exception != null) {
+                exceptions.add(exception.getMessage());
+            }
+        }
+
+        if (exceptions.isEmpty() == false) {
+            throw new IllegalStateException(String.join(", ", exceptions));
+        }
+
+        return requests;
+    }
+
+    @Override
+    public TransformState transform(Object source, TransformState prevState) throws Exception {
+        // We execute the prepare() call to catch any errors in the transform phase
+        prepare(source);
+        return prevState;
+    }
+
+    @Override
+    public Set<String> postTransform(Object source, ClusterState clusterState, Set<String> previousKeys) {
+        var requests = prepare(source);
+
+        for (var request : requests) {
+            roleMappingStore.putRoleMapping(request, new ActionListener<>() {
+                @Override
+                public void onResponse(Boolean aBoolean) {}
+
+                @Override
+                public void onFailure(Exception e) {
+                    throw new IllegalStateException("Error creating role mapping [" + request.getName() + "]", e);
+                }
+            });
+        }
+
+        Set<String> entities = requests.stream().map(r -> r.getName()).collect(Collectors.toSet());
+
+        Set<String> toDelete = new HashSet<>(previousKeys);
+        toDelete.removeAll(entities);
+
+        for (var mappingToDelete : toDelete) {
+            var deleteRequest = new DeleteRoleMappingRequest();
+            deleteRequest.setName(mappingToDelete);
+
+            roleMappingStore.deleteRoleMapping(deleteRequest, new ActionListener<>() {
+                @Override
+                public void onResponse(Boolean aBoolean) {}
+
+                @Override
+                public void onFailure(Exception e) {
+                    throw new IllegalStateException("Error deleting role mapping [" + mappingToDelete + "]", e);
+                }
+            });
+        }
+
+        return Collections.unmodifiableSet(entities);
```

Comment:
Deleting role mapping (as well as updating) is async. So they could fail even when the return statement successfully executed. This means ES could report it successfully updated the file-based-settings while the role mapping persistence could fail? 

I understand that we accept that cluster and index operations are separate and can fail independently. But I think it should be: "When the overall process is reported as failure, there can be part of the update that have been successful."

But **not** "When the overall process is reported as success, there can be part of the update that have failed." Because this would prevent the caller from retrying on failure and leave the system in an inconsistent state indefinitely.

Concretely, I am suggesting that we should handle the async errors (inside listeners) before returning.

Reply:
You are right, this is a bug. I should be waiting for the async actions to complete.